M1 Mac support

Make `nix build` (and `bin/run`) succeed on M1 Mac with native Nix installed.

No rosetta.

- [x] `nix build` works
- [x] `nix develop` works, with HLS, ormolu, etc.
- [ ] GitHub CI (pending on https://github.com/actions/virtual-environments/issues/2187 )

`/nix/store/7vjmcgcxm0jy0w2jsyk99n0kiq52q296-compiler-rt-libc-9.0.1.drv` failed to build:

```
[ 48%] Building C object lib/builtins/CMakeFiles/clang_rt.cc_kext_arm64_osx.dir/floatsidf.c.o
/tmp/nix-build-compiler-rt-libc-9.0.1.drv-0/compiler-rt-9.0.1.src/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:168:31: error: invalid application of 'sizeof' to an incomplete
  unsigned struct_stat64_sz = sizeof(struct stat64);
                              ^     ~~~~~~~~~~~~~~~
/tmp/nix-build-compiler-rt-libc-9.0.1.drv-0/compiler-rt-9.0.1.src/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:168:45: note: forward declaration of '__sanitizer::stat64'
  unsigned struct_stat64_sz = sizeof(struct stat64);
                                            ^
/tmp/nix-build-compiler-rt-libc-9.0.1.drv-0/compiler-rt-9.0.1.src/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:194:33: error: invalid application of 'sizeof' to an incomplete
  unsigned struct_statfs64_sz = sizeof(struct statfs64);
                                ^     ~~~~~~~~~~~~~~~~~
/tmp/nix-build-compiler-rt-libc-9.0.1.drv-0/compiler-rt-9.0.1.src/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:194:47: note: forward declaration of '__sanitizer::statfs64'
  unsigned struct_statfs64_sz = sizeof(struct statfs64);
                                              ^
[ 49%] Building C object lib/builtins/CMakeFiles/clang_rt.builtins_arm64_osx.dir/fixunsdfdi.c.o
2 errors generated.
make[2]: *** [lib/sanitizer_common/CMakeFiles/RTSanitizerCommonNoHooks.osx.dir/build.make:370: lib/sanitizer_common/CMakeFiles/RTSanitizerCommonNoHooks.osx.dir/sanitizer_platform_limits_
make[1]: *** [CMakeFiles/Makefile2:851: lib/sanitizer_common/CMakeFiles/RTSanitizerCommonNoHooks.osx.dir/all] Error 2
```

--- a/flake.nix
+++ b/flake.nix
@@ -14,6 +14,21 @@
         overlays = [ ];
         pkgs =
           import nixpkgs { inherit system overlays; config.allowBroken = true; };
+        # https://github.com/NixOS/nixpkgs/issues/140774#issuecomment-976899227
+        m1MacHsBuildTools =
+          pkgs.haskellPackages.override {
+            overrides = self: super:
+              let
+                workaround140774 = hpkg: with pkgs.haskell.lib;
+                  overrideCabal hpkg (drv: {
+                    enableSeparateBinOutput = false;
+                  });
+              in
+              {
+                ghcid = workaround140774 super.ghcid;
+                ormolu = workaround140774 super.ormolu;
+              };
+          };
         project = returnShellEnv:
           pkgs.haskellPackages.developPackage {
             inherit returnShellEnv;
@@ -25,16 +40,18 @@
               # cf. https://tek.brick.do/K3VXJd8mEKO7
             };
             modifier = drv:
-              pkgs.haskell.lib.addBuildTools drv (with pkgs.haskellPackages;
-              [
-                # Specify your build/dev dependencies here. 
-                cabal-fmt
-                cabal-install
-                ghcid
-                haskell-language-server
-                ormolu
-                pkgs.nixpkgs-fmt
-              ]);
+              pkgs.haskell.lib.addBuildTools drv
+                (with (if system == "aarch64-darwin"
+                then m1MacHsBuildTools
+                else pkgs.haskellPackages); [
+                  # Specify your build/dev dependencies here. 
+                  cabal-fmt
+                  cabal-install
+                  ghcid
+                  haskell-language-server
+                  ormolu
+                  pkgs.nixpkgs-fmt
+                ]);
           };
       in
       {

